src/components/chat/ui/ChatMessageList.tsx

```diff
@@ -0,0 +1,142 @@
+import { css } from '@linaria/core';
+import { isSameDay } from 'date-fns/isSameDay';
+
+import { getComponentKeyFromMessage } from '@uikit/modules/GroupChannel/context/utils';
+import { isSendableMessage } from '@uikit/utils';
+
+import { useConstantState } from '../../../context/ConstantContext';
+import { DateSeparator } from '../../../foundation/components/DateSeparator';
+import FrozenBanner from '../../../foundation/components/FrozenBanner';
+import { InfiniteMessageList } from '../../../foundation/components/InfiniteMessageList';
+import { Placeholder } from '../../../foundation/components/Placeholder';
+import { ScrollToBottomButton } from '../../../foundation/components/ScrollToBottomButton';
+import { isDashboardPreview } from '../../../utils';
+import { getMessageGrouping } from '../../../utils/messages';
+import CustomMessage from '../../CustomMessage';
+import MessageDataContent from '../../MessageDataContent';
+import SuggestedRepliesContainer from '../../SuggestedRepliesContainer';
+import { useChatContext } from '../context/ChatProvider';
+import { useBotStudioView } from '../hooks/useBotStudioView';
+import { useTypingTargetMessageId } from '../hooks/useTypingTargetMessageId';
+
+export const ChatMessageList = () => {
+  const { channel, dataSource, scrollSource, handlers } = useChatContext();
+  const { botStudioEditProps, customUserAgentParam, stringSet } = useConstantState();
+
+  const typingTargetMessageId = useTypingTargetMessageId();
+  const { filteredMessages, shouldShowOriginalDate, renderBotStudioWelcomeMessages } = useBotStudioView();
+
+  const render = () => {
+    if (!dataSource.initialized) {
+      return <Placeholder type={'loading'} />;
+    }
+
+    if (dataSource.messages.length === 0) {
+      return <Placeholder type={'noMessages'} />;
+    }
+
```

Comment:
minor

```
const { scrollRef, scrollPositionRef, scrollDistanceFromBottomRef, setIsScrollBottomReached } = scrollSource;
```

Reply:
The logic in this scope is quite complex, and having many local variables could lead to many variables being selected in the type system. Therefore, destructuring was not used.